neuralmonkey/attention/base_attention.py

```diff
@@ -0,0 +1,282 @@
+            name="contexts"),
+        weights=tf.TensorArray(
+            dtype=tf.float32, size=0, dynamic_size=True,
+            name="distributions"))
+
+
+def get_attention_states(encoder: Union[TemporalStateful,
+                                        SpatialStateful]) -> tf.Tensor:
+    if isinstance(encoder, TemporalStateful):
+        return encoder.temporal_states
+
+    elif isinstance(encoder, SpatialStateful):
+        # pylint: disable=no-member
+        shape = [s.value for s in encoder.spatial_states.get_shape()[1:]]
+        # pylint: enable=no-member
+        return tf.reshape(encoder.spatial_states,
+                          [-1, shape[0] * shape[1], shape[2]])
+    else:
+        raise AssertionError("Unknown encoder type")
+
+
+def get_attention_mask(encoder: Union[TemporalStateful,
+                                      SpatialStateful]) -> Optional[tf.Tensor]:
+    if isinstance(encoder, TemporalStateful):
+        return encoder.temporal_mask
+
+    elif isinstance(encoder, SpatialStateful):
+        if encoder.spatial_mask is None:
+            return None
+
+        # pylint: disable=no-member
+        shape = [s.value for s in encoder.spatial_mask.get_shape()[1:]]
+        # pylint: enable=no-member
+        return tf.reshape(encoder.spatial_mask, [-1, shape[0] * shape[1]])
+    else:
+        raise AssertionError("Unknown encoder type")
+
+
+class BaseAttention(ModelPart):
+    def __init__(self,
+                 name: str,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+
+        self.query_state_size = None  # type: tf.Tensor
+        self._histories = {}  # type: Dict[str, tf.Tensor]
+
+    @property
+    def histories(self) -> Dict[str, tf.Tensor]:
+        return self._histories
+
+    def attention(self,
+                  query: tf.Tensor,
+                  decoder_prev_state: tf.Tensor,
+                  decoder_input: tf.Tensor,
+                  loop_state: Any,
+                  step: tf.Tensor) -> Tuple[tf.Tensor, Any]:
+        """Get context vector for a given query."""
+        raise NotImplementedError("Abstract method")
+
+    def initial_loop_state(self) -> Any:
+        """Get initial loop state for the attention object."""
+        raise NotImplementedError("Abstract method")
+
+    def finalize_loop(self, key: str, last_loop_state: Any) -> None:
+        raise NotImplementedError("Abstract method")
+
+    # pylint: disable=no-self-use
+    @tensor
+    def train_mode(self) -> tf.Tensor:
+        return tf.placeholder(tf.bool, [], "train_mode")
+    # pylint: enable=no-self-use
+
+    def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
+        return {self.train_mode: train}
+
+    @property
+    def context_vector_size(self) -> int:
+        raise NotImplementedError("Abstract property")
+
+
+class Attention(BaseAttention):
```

Comment:
Možná bych to přestěhoval do samostatného souboru a udělal znova ten ten trik s `__init__`. Pak to chce přidat docstring, kterej řekne, že tohle je "ten" attention z Bahdanaua.